test: Fix test by checking the actual exception instance

hebasto opened the pull request:
The `system_tests/run_command` test is broken because it passes even with the diff as follows:
```diff
--- a/src/test/system_tests.cpp
+++ b/src/test/system_tests.cpp
@@ -90,7 +90,7 @@ BOOST_AUTO_TEST_CASE(run_command)
         });
     }
     {
-        BOOST_REQUIRE_THROW(RunCommandParseJSON("echo \"{\""), std::runtime_error); // Unable to parse JSON
+        BOOST_REQUIRE_THROW(RunCommandParseJSON("invalid_command \"{\""), std::runtime_error); // Unable to parse JSON
     }
     // Test std::in, except for Windows
 #ifndef WIN32
```

The reason of such fragility is that the [`BOOST_REQUIRE_THROW`](https://www.boost.org/doc/libs/1_83_0/libs/test/doc/html/boost_test/utf_reference/testing_tool_ref/assertion_boost_level_throw.html) macro passes even if the command raises an exception in the underlying subprocess implementation, which might have a type derived from `std::runtime_error`.

furszy commented on src/test/system_tests.cpp:
```diff
@@ -90,7 +92,15 @@ BOOST_AUTO_TEST_CASE(run_command)
+#else
+        const std::string command{"echo {"};
+#endif
+        BOOST_CHECK_EXCEPTION(RunCommandParseJSON(command), std::runtime_error, [&](const auto& e) {
+            return std::strstr(e.what(), "Unable to parse JSON: {") != nullptr;
+        });
```
Could use `HasReason()`.
```c++
BOOST_CHECK_EXCEPTION(RunCommandParseJSON(command), std::runtime_error, HasReason("Unable to parse JSON"));
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/28989#issuecomment-1837526938).

hebasto committed:
test: Fix test by checking the actual exception instance

The BOOST_REQUIRE_THROW passes even if the command raises an exception
in the underlying subprocess implementation, which might have a type
derived from std::runtime_error.

furszy reviewed (commented):
Non-Windows code ACK 55e3dc3e

pablomartin4btc reviewed (commented):
ACK 55e3dc3e03510e97caba1547a82e3e022b0bbd42

hebasto commented:
Friendly ping @Sjors :)
